ateam_ai/src/trajectory_generation/b_spline.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright 2021 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "trajectory_generation/b_spline.hpp"
+#include <iostream>
+namespace BSpline {
+std::vector<Eigen::Vector2d> apply_multiplicity(const std::vector<Eigen::Vector2d> & control_points, std::size_t multiplicity) {
+  std::vector<Eigen::Vector2d> out;
+  for (int i = 0; i < multiplicity; i++) {
+    out.push_back(control_points.front());
+  }
+  for (int i = 1; i < control_points.size() - 1; i++) {
+    out.push_back(control_points.at(i));
+  }
+  for (int i = 0; i < multiplicity; i++) {
+    out.push_back(control_points.back());
+  }
+
+  return out;
```

Comment:
Consider an algorithm-based implementation
```suggestion
  std::vector<Eigen::Vector2d> out(control_points);
  std::fill_n(std::front_inserter(out), multiplicity, control_points.front());
  std::fill_n(std::back_inserter(out), multiplicity, control_points.front());
  return out;
```

Reply:
Rip, std::front_inserter requires push_front that vector doesn't have. I'll add a todo for later and leave it as is for now. The code is no where near performant as is